### ISSUE TITLE
`Number.toString(base)` does not have `@nosideeffects` semantics

### ISSUE THREAD

**Issue by mnsaglam** [google/closure-compiler#4169]
It is mentioned in various places that `.toString()` method has a hardcoded `@nosideeffects` property.  (e.g., [Angular nosideeffect annotation)](https://github.com/angular/angular/blob/67b2c336bc0bdce3f7ae054c094990a9831f5b20/packages/core/src/util/closure.ts#L9-L20
)) (On a side note, is this hack still relevant after the @nosideeffects being allowed in non-externs code?)

However the same is not true for `Number.toString(base)` variant, which writes the number in bases other than base 10.

```javascript
// a.js
for (let i = 0; i < 10; ++i)
  i.toString();
for (let j = 0; j < 10; ++j)
  j.toString(16);
```

```shell
google-closure-compiler -O ADVANCED a.js

a.js:2:2: WARNING - [JSC_USELESS_CODE] Suspicious code. This code lacks side-effects. Is there a bug?
  2|   i.toString();
       ^^^^^^^^^^^^

0 error(s), 1 warning(s), 94.7% typed
```

The first one is detected as suspicious code, the second one isn't.

The relevant  [externs definition](https://github.com/google/closure-compiler/blob/573aad7b4f1f7f9320b8d52a0ec2831b2b468ea7/externs/es3.js#L1191-L1200) appears correct, however it looks like there some special handling of `.toString()` in the java code.

This is important because in many fields (crypto, verifiable computation) some constants need to be generated (instead of being written out as literals). If GCC can eliminate the ones that aren't used, it will result in significant reductions. However when the generation code uses `toString(16)` elimination does not happen.

**Comment by brad4d** [comment]
We've added this issue to an internal list of possible optimization improvements, but it isn't a high priority.

We would likely accept a pull request.

**Comment by MadhuriRathod30** [comment]
Can I take this up ? Can you please assign this to me ?

**Comment by brad4d** [comment]
@MadhuriRathod30 

Well, it looks like I can assign it to you, which actually surprised me a bit.
Earlier today I was unable to assign a different issue to one of our team members, because we hadn't put them into the appropriate GitHub team yet.

I cannot explain that, but you're welcome. :)

**Comment by MadhuriRathod30** [comment]
@brad4d Thank you :-)
